### DISCUSSION TITLE
A very practical and attractive idea: map compression

### DISCUSSION THREAD

**Discussion by aaa1386** [freeplane/freeplane#1964] [Ideas]
To better understand the importance of this idea, **carefully read the page [https://github.com/freeplane/freeplane/discussions/1936](https://github.com/freeplane/freeplane/discussions/1936).** This page explains the significance of viewing nodes together, especially sibling, child, and uncle nodes.

Look at the image below:  
In many cases, users have to fold nodes or scroll the map to view them. This process is time-consuming and reduces the user experience. **Why not utilize the empty spaces present in the map?**  

In this example, the user has created many nodes, but even in typical cases, the map generates considerable empty space.  

If a **"Map Compression" option** existed, accessing nodes would become significantly easier. This feature would enable users to make the most of unused spaces and view more nodes simultaneously.  
You wouldn't need a larger screen—this option would expand the working area within Freeplane for you. Imagine seeing **30 nodes instead of just 10 on the same screen.** Wouldn't that be interesting?  

![image](https://github.com/user-attachments/assets/e7a99861-30ab-41a2-b2ab-9833d5d0ff44)

---

In the image below, you can see the **unnecessary distance between nodes.** These gaps not only make navigation difficult but also detract from the map's visual appeal. **If there were an option to reduce these distances, it would be invaluable.**  

![image](https://github.com/user-attachments/assets/1852c2e6-8df9-4485-9c23-e88cc889c26c)

![image](https://github.com/user-attachments/assets/3fc36cb0-e8f5-46fd-9051-ef344b235425)

---

In this example, **the siblings of Node 1 are not visible.** Nodes 3 and 4, which are siblings, are too far apart. This unnecessary spacing not only complicates access but also diminishes the aesthetic of the map.  

The idea of compressing the map is both practical and visually appealing.  

I don't need to explain much further—**the importance of this idea is self-evident to any experienced user.** However, if more clarification is needed, feel free to ask for additional explanations.  

---

In the image below, I've highlighted **the unused spaces that can be removed** using rectangles. Removing these empty spaces makes the map more optimized and improves node accessibility.  

![image](https://github.com/user-attachments/assets/8c7d08a3-192d-4feb-ac2e-3632d0f29a8f)

The following image show the map **before and after compression:**  

![Image](https://github.com/user-attachments/assets/fba82cb5-4510-4040-b066-eb033514c531)

### The Idea: Map Compression

A new feature called **"Map Compression"** should be added to Freeplane. This feature removes unnecessary gaps, making the map more compact and increasing the number of visible nodes on the screen.  

#### Key Benefits:  
1. **Optimized for Small Screens:** This feature is particularly helpful for users working with small displays.  
2. **Enhanced Focus:** By compressing the map, the need to fold nodes decreases, and users can view more nodes at once.  
3. **Simplified Navigation:** Closer spacing between nodes makes it easier to navigate the map.  

---

### Preserving Edge Structures

The compression process **does not alter the initial structure of edges.** A small horizontal line can be added from the end of the node to the start of the edge division. This adjustment prevents edges from overlapping with nodes below them.  

Refer to the image below: for compression, edges need to start slightly farther from the nodes. I've marked the required adjustment with a **red line.**  

**Before Compression:**  
![image](https://github.com/user-attachments/assets/f36f6a4f-9666-47f6-a446-3f5a53561f60)  

**After Compression:**  
![image](https://github.com/user-attachments/assets/0f49a5c4-be5c-455e-88d4-87333877c9f0)

---

### Dynamic Adaptation  

When users fold or unfold nodes, **Freeplane should automatically adapt the map to the new structure.**  
Additionally, this feature should be optional, allowing users to enable or disable map compression at any time.

**Comment by aaa1386** [comment]
If this idea is implemented, I will propose another good idea that makes this idea more complete

**Comment by aaa1386** [comment]
This idea creates a fundamental transformation in FP.
Among the ideas I gave for FP, I like this idea the most. With this idea, communication with FP becomes very easy. It is very useful especially for those who have a small screen.
Thanks to the creator of FP and his attention to ideas.

**Comment by quickfold** [comment]
I love this idea. The main reason I continue to use Freeplane is its ability to display more ideas on screen at once than text editors. This idea would make it even better at that task.

**Comment by Beryesa** [comment]
This would be **really** useful on navigating busy maps or even the PNG exports

**Comment by euu2021** [comment]
The idea is interesting! I also like FP because of its information compression capabilities.

Now, how would it work in a map with nodes of varied width? For example, in this situation:


![image](https://github.com/user-attachments/assets/42693d34-0a0e-465b-bf8d-754526b1efa5)

The `dddddddd` node doesn't fit in the gap:

![image](https://github.com/user-attachments/assets/48db91a5-b55c-4220-b234-649bd64e0c6a)

**Reply by aaa1386** [reply]
This idea is not for such cases. Wherever the map can be compressed, it will be compressed. If some nodes do not have that capability, they will not be compressed.
The edges of the "CCC" nodes should be pulled to the right to create space for the "DDD" node.   Of course, it can be done to some extent, not much.

**Comment by euu2021** [comment]
It's interesting that, in some situations, there would be less jumping around of the map. The jumping around of the map is a bit confusing to the eyes.

For example, in this situation, when the user fold/unfold node ccc, there is, currently, a general jumping around of nodes (hhh, iii, bbb, ddd):

![javaw_khLL0p8HYb](https://github.com/user-attachments/assets/2af41d41-5dca-4bf6-91d9-32953426fd6f)

On the other hand, in map compression mode, there those jumps wouldn't happen:

![mspaint_8kaT39lFs0](https://github.com/user-attachments/assets/0917ab3d-9b16-4dfc-88e6-05348f379730)


Of course, the jumps would still happen in other situations.

**Comment by dpolivaev** [comment]
The idea is interesting and quite difficult to implement. At the moment the tags still have my main attention. Maybe I come back to this idea later after other issues are solved.

**Reply by aaa1386** [reply]
Thanks to the FP developer.
If this idea is implemented, it will be very practical and useful. Those who work with FP a lot understand this very well.
I hope someday this idea will be created in FP.

**Comment by guoranjun2** [comment]
I was just thinking about this yesterday, and I really think this idea is great.

![image](https://github.com/user-attachments/assets/cd0bb05d-bda5-4d43-8d76-33a39d868ac4)

![image](https://github.com/user-attachments/assets/05302205-d49e-4667-9db2-a77c08954051)

**Comment by Lotus49b** [comment]
An interesting idea. 

Personally I find that the Map Overview gives me the "big picture" view that I need. It's also a good idea to create templates that give you the best layout for the type of work you are doing. 
![image](https://github.com/user-attachments/assets/69fc9d26-5658-4b4a-a1ca-c6af6fb2e70c)

**Comment by dpolivaev** [comment]
Implemented in preview 1.12.11_01
See https://github.com/freeplane/freeplane/discussions/2409

**Comment by aaa1386** [comment]
Thank you very much for the effort you put into implementing this idea.
I had a question, what does the second option you see in the image do?

![image](https://github.com/user-attachments/assets/f7f5f2c2-b5fb-47bf-8d88-a0d48fed2852)

**Reply by euu2021** [reply]
> I had a question, what does the second option you see in the image do?

From https://github.com/freeplane/freeplane/discussions/379#discussion-4083417:

> Since version 1.10.3_02 there is also a new layout option View->View settings->Allow compact layout. It can be activated per map and saved in a map template. If it is enabled, moving nodes up allows to overlap them with any nodes above making creating user specific layouts easier than using the free positioned nodes.

**Reply by aaa1386** [reply]
Wouldn't it be better to call it:  "Allow Overlapping of Nodes When Manually Repositioned"
The current title is confusing and may be confused with the new option.
